libutils/json-utils.c

```diff
@@ -0,0 +1,263 @@
+    {
+        return;
+    }
+
+    const char *myname = "ParseEnvLine";
+    int line_length = strlen(line);
+
+    if (line[0] == '#' || line_length == 0)
+    {
+        return;
+    }
+    char *next_equal = strchr(line, '=');
+    if (next_equal == NULL)
+    {
+        Log(LOG_LEVEL_DEBUG, "%s: The line '%s' in ENV file '%s' isn't empty, but was skipped because it's missing an equal sign",
+            myname, line, filename_for_log);
+        return;
+    }
+    int equal_index = (int)(next_equal - line);
+    if (equal_index == 0)
+    {
+        Log(LOG_LEVEL_DEBUG, "%s: The line '%s' in ENV file '%s' was skipped because it's missing a key",
+            myname, line, filename_for_log);
+        return;
+    }
+    *next_equal = '\0';
+    key = TrimWhitespace(line);
+    value = TrimWhitespace(next_equal + 1);
+    value = filtered_copy(value, value);
+
+    *key_out = key;
+    *value_out = value;
+}
+
+/**
+ * @brief Opens an env file, prints an error message on failure
+ *
+ * @param input_path Path to env file
+ * @param myname Descriptive name of caller for logging purposes
+ */
+static FILE *ParseEnvOpenFile(const char *input_path, const char *myname)
+{
+    assert(input_path);
+    assert(myname);
+    FILE *fin = safe_fopen(input_path, "r");
+    if (fin == NULL)
+    {
+        Log(LOG_LEVEL_VERBOSE, "%s cannot open the ENV file '%s' (fopen: %s)",
+            myname, input_path, GetErrorStr());
+
+    }
+    return fin;
+}
+
+static void ByteLimitLog(const char *myname, int linenumber, const char *input_path, size_t size_max)
+{
+    Log(LOG_LEVEL_VERBOSE, "%s: line %d from ENV file '%s' exceeded byte limit %zu, done with file",
+        myname, linenumber, input_path, size_max);
+}
+
+/**
+ * @brief Parses an env file and creates a key-value pair json element
+ *
+ * Creates JSON element where all keys and values are strings
+ *
+ * @param input_path file to read from ex: "/etc/os-release"
+ * @param size_max   Maximum size of env file (in bytes)
+ * @param json_out   Where to save pointer to new JsonElement, must destroy
+ * @return 0 = success
+ */
+int JsonParseEnvFile(const char *input_path, size_t size_max, JsonElement **json_out)
+{
+    assert(json_out);
+    *json_out = JsonObjectCreate(10);
+    const char *myname = "JsonParseEnvFile";
+    size_t line_size = CF_BUFSIZE;
+    char *raw_line = xmalloc(CF_BUFSIZE);
+    char *key, *value;
+    int linenumber = 0;
+    size_t byte_count = 0;
+
+    FILE *fin = ParseEnvOpenFile(input_path, myname);
+    if (fin == NULL)
+    {
+        return -1;
+    }
+
+    while (CfReadLine(&raw_line, &line_size, fin) != -1)
+    {
+        ++linenumber;
+        int line_length = strlen(raw_line);
+        byte_count += line_length;
+        if (byte_count > size_max)
+        {
+            ByteLimitLog(myname, linenumber, input_path, size_max);
+            break;
+        }
+
+        ParseEnvLine(raw_line, &key, &value, input_path);
+        if (key != NULL && value != NULL)
```

Comment:
Added some log messages to ParseEnvLine, for invalid syntax, missing key, etc. Will not log on all cases of NULL - I don't want to fill the log with debug messages about empty lines or commented out lines.